service/test/test_write_session.sh

```diff
@@ -0,0 +1,109 @@
+#!/bin/bash
+#  Copyright (c) 2015 - 2021, Intel Corporation
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions
+#  are met:
+#
+#      * Redistributions of source code must retain the above copyright
+#        notice, this list of conditions and the following disclaimer.
+#
+#      * Redistributions in binary form must reproduce the above copyright
+#        notice, this list of conditions and the following disclaimer in
+#        the documentation and/or other materials provided with the
+#        distribution.
+#
+#      * Neither the name of Intel Corporation nor the names of its
+#        contributors may be used to endorse or promote products derived
+#        from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+#  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+#  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+set -x
+
+# This test shows the GEOPM service being used to set the maximum CPU
+# frequency of core zero to 2 GHz.  The test uses the geopmsession
+# command line tool to read and write from/to the MSR_PERF_CTL
+# register to control the maximum frequency of the core.  The test
+# first reads the value of the register, then opens a write session to
+# set it to 2 GHz and reads the register again.  The test then kills the
+# write session with signal 9 and reads the control register a third
+# time.  The test asserts that the control value was changed by the
+# write sesssion, and that this change was reverted to the value it
+# started with after the session is killed.
+
+# PARAMETERS
+CONTROL=MSR::PERF_CTL:FREQ
+DOMAIN=core
+DOMAIN_IDX=0
+CONTROL_VALUE=2000000000.0
+REQUEST="${CONTROL} ${DOMAIN} ${DOMAIN_IDX}"
+SESSION_DIR=/var/run/geopm-service
+
+# READ CONTROL REGISTER
+START_VALUE=$(echo ${REQUEST} | geopmsession)
```

Comment:
I think that if the service is not running, then the initial `geopmsession` call should fail.  In that case, we should print a nice error that the service isn't running if we don't already somewhere.

Reply:
```
systemctl status geopm
● geopm.service - Global Extensible Open Power Manager Service
   Loaded: loaded (/usr/lib/systemd/system/geopm.service; disabled; vendor preset: disabled)
   Active: inactive (dead)
cmcantal@mcfly:~/Git/geopm/service (geopm-service)]$ ./test/test_write_session.sh 
+ CONTROL=MSR::PERF_CTL:FREQ
+ DOMAIN=core
+ DOMAIN_IDX=0
+ CONTROL_VALUE=2000000000.0
+ REQUEST='MSR::PERF_CTL:FREQ core 0'
+ SESSION_DIR=/var/run/geopm-service
++ echo MSR::PERF_CTL:FREQ core 0
++ geopmsession
Traceback (most recent call last):
  File "/usr/bin/geopmsession", line 37, in <module>
    main()
  File "/usr/lib/python3.6/site-packages/geopmdpy/session.py", line 118, in main
    read_session(geopm_proxy, args.time, args.period, requests)
  File "/usr/lib/python3.6/site-packages/geopmdpy/session.py", line 41, in read_session
    geopm_proxy.PlatformOpenSession()
  File "/usr/lib/python3.6/site-packages/dasbus/client/proxy.py", line 161, in __getattr__
    member = self._get_member(self._get_interface(name), name)
  File "/usr/lib/python3.6/site-packages/dasbus/client/proxy.py", line 211, in _get_interface
    self._handler.specification.members
  File "/usr/lib/python3.6/site-packages/dasbus/client/handler.py", line 206, in specification
    self._specification = self._get_specification()
  File "/usr/lib/python3.6/site-packages/dasbus/client/handler.py", line 330, in _get_specification
    "(s)"
  File "/usr/lib/python3.6/site-packages/dasbus/client/handler.py", line 447, in _call_method
    **kwargs,
  File "/usr/lib/python3.6/site-packages/dasbus/client/handler.py", line 477, in _get_method_reply
    return self._handle_method_error(error)
  File "/usr/lib/python3.6/site-packages/dasbus/client/handler.py", line 497, in _handle_method_error
    raise exception from None
dasbus.error.DBusError: The name io.github.geopm was not provided by any .service files
+ START_VALUE=
```

That is the current output.  I put up a todo about how command line utilities should do exception handling so the error message is cleaner, but it is not bad as is from DBus.  Could easily put a try: except: in the geopmdpy.session.main() and print the exception error string rather than the full backtrace in a future patch.